feat: instrument multi-copy upload outcomes (#501)
filecoin-project/filecoin-pin: README.md

* With our [end user affordance](#affordances) we expect to make telemetry on by default, requiring a consumer/user to opt out.  We are defaulting as "enabled" to help make sure we have a good pulse on the user experience and can address issues correctly.

> * In this [pre-v1 season](https://github.com/filecoin-project/filecoin-pin/issues/187), we are particularly focused on helping maintainers validate functionality and iron out problems throughout the whole Filecoin Onchain Cloud stack that `filecoin-pin` relies on.

Reviewer: ```suggestion
## Telemetry

Filecoin Pin collects telemetry.  A few things:
* Telemetry always [has a way to be disabled](#how-to-disable-telemetry).
* We don't collect Personal identifiable information (PII).
* [Affordance](#affordances) make telemetry on by default, requiring a consumer/user to opt out.  We are defaulting as "enabled" to help make sure we have a good pulse on the user experience and can address issues correctly. Maintainers are particularly focused on validating functionality and ironing out problems throughout the whole Filecoin Onchain Cloud stack that `filecoin-pin` relies on.
```

What we collect:
* **Error reports** via Sentry (no PII).
* **Per-upload copy outcomes** posted directly to [BetterStack's HTTP metrics ingestion endpoint](https://betterstack.com/docs/logs/ingesting-data/http/metrics/), so we can measure the success rate of multi-copy uploads and identify which storage providers (or pipeline steps) are failing. Two metrics are emitted per resolved copy:
  * `uploadCopyStatus` counter — one increment per resolved copy attempt. Tags: `spId` (storage provider ID), `role` (`primary`/`secondary`), `value` (`success`, `failure.pull`, `failure.commit`, `failure.other`), `network`, `affordance` (which surface emitted the metric: `CLI`, `GitHub Action`, `Library`, `Filecoin Pin Website`).
  * `uploadCopySize` gauge — the piece size in bytes for the same outcome, sharing the counter's tag set so dashboards can join the two (e.g. p99 size of commit-step failures).

>   See [`documentation/events-and-metrics.md`](documentation/events-and-metrics.md) for the full schema, including the underlying events and the relationship between this metric and the Synapse SDK's upload result.

Reviewer: ```suggestion
* **Per-upload copy outcomes** posted directly to [BetterStack's HTTP metrics ingestion endpoint](https://betterstack.com/docs/logs/ingesting-data/http/metrics/), so we can measure the success rate of multi-copy uploads and identify which storage providers (or pipeline steps) are failing. See [`documentation/events-and-metrics.md`](documentation/events-and-metrics.md) for the full schema, including the underlying events and the relationship between this metric and the Synapse SDK's upload result.
```

Have one place where we're documenting the specifics...

>   **Delivery model.** Each `executeUpload` fires its own HTTP POST containing one counter point per copy and per failed attempt — there is no in-memory buffer or periodic flush. The CLI, pinning server, and GitHub Action `await flushTelemetry()` before exit so any in-flight request finishes. **Long-running consumers that terminate via `process.exit()`, `SIGINT`, or `SIGTERM` should do the same** (`flushTelemetry` is exported from `filecoin-pin/core/telemetry`). To silence subsequent `recordUploadResult` calls without exiting the process, call `configureTelemetry({ disabled: true })`.

Reviewer: Head is feeling blurry, but I was struggling a little on reasoning why these specific details about executeUpload are in the readme vs. events-and-metrics.md. 

I don't have strong convictions here...

Author: Events and metrics seems more like a data type definition, while this section is relevnt for library users wanting to ensure all telemetry is submitted

* In this [pre-v1 season](https://github.com/filecoin-project/filecoin-pin/issues/187), we are particularly focused on helping maintainers validate functionality and iron out problems throughout the whole Filecoin Onchain Cloud stack that `filecoin-pin` relies on.

What we collect:
* **Error reports** via Sentry (no PII).
* **Per-upload copy outcomes** posted directly to [BetterStack's HTTP metrics ingestion endpoint](https://betterstack.com/docs/logs/ingesting-data/http/metrics/), so we can measure the success rate of multi-copy uploads and identify which storage providers (or pipeline steps) are failing. One counter is emitted:
  * `uploadCopyStatus` — incremented once per resolved copy attempt. Tags: `spId` (storage provider ID), `role` (`primary`/`secondary`), `value` (`success`, `failure.pull`, `failure.commit`, `failure.other`), `network`, `affordance` (which surface emitted the metric: `CLI`, `GitHub Action`, `Library`, `Filecoin Pin Website`).

  See [`documentation/events-and-metrics.md`](documentation/events-and-metrics.md) for the full schema, including the underlying events and the relationship between this metric and the Synapse SDK's upload result.

>   The BetterStack source token is shipped in source, so the data source is treated as public and untrusted. The CLI lets you send your own copy with `FILECOIN_PIN_METRICS_ENDPOINT` and `FILECOIN_PIN_METRICS_TOKEN`.

Reviewer 2: idk if we should have this in the README
```suggestion
```

  **Library usage (Node and browser).** The telemetry library never reads `process.env`. Configure it programmatically before the first `executeUpload` — the same API works in both runtimes:

  ```ts
  import { configureTelemetry } from 'filecoin-pin/core/telemetry'

  configureTelemetry({ disabled: true })                       // opt out
  configureTelemetry({ endpoint: '…', token: '…' })            // send elsewhere
  configureTelemetry({ affordance: 'Filecoin Pin Website' })   // tag the surface (default 'Library')
  ```

  The CLI's env-var support is built on top of this API (see `src/read-telemetry-config-from-env.ts`); other Node hosts can follow the same pattern.

### How to disable telemetry

- **CLI / pinning server:** set `FILECOIN_PIN_TELEMETRY_DISABLED=true` (or the cross-tool standard `DO_NOT_TRACK=1`).

> - **GitHub Action:** pass `disableTelemetry: true` to the action (see [`upload-action/action.yml`](upload-action/action.yml)).

Reviewer 2: we should support the same env vars as CLI here. I'm fine with leaving the action input as well..

Author: Can you explain why? I don't want to duplicate all settings when the primary configuration method of action inputs works just fine

Author: will add support, ChatGPT convinced: https://chatgpt.com/s/t_6a18286aa2288191ba021e2dedf33f59

Author: 019ddc4

DATABASE_PATH=./pins.db        # SQLite database location
CAR_STORAGE_PATH=./cars        # CAR file storage directory
LOG_LEVEL=info                 # Logging verbosity (info, debug, error)

> # Optional - Telemetry (see "CLI Telemetry" above)

Reviewer: ```suggestion
# Optional - Telemetry (see "Telemetry" above)
```

FILECOIN_PIN_TELEMETRY_DISABLED=true        # Disable all telemetry
DO_NOT_TRACK=1                              # Standard cross-tool opt-out
FILECOIN_PIN_METRICS_ENDPOINT=https://…     # Override BetterStack metrics endpoint

> FILECOIN_PIN_METRICS_TOKEN=…                # Override BetterStack source token

Reviewer: Are there options for the "Sentry" metrics endpoint?

Author: No. I'm going to remove this configuratino, I don't see us needing it

Author: b8b5b5d
